src/H5Tconv.c

```diff
@@ -3024,23 +3066,21 @@ H5T__conv_enum_numeric(hid_t src_id, hid_t dst_id, H5T_cdata_t *cdata, size_t ne
             break;
 
         case H5T_CONV_CONV:
-            if (NULL == (src = (H5T_t *)H5I_object(src_id)) || NULL == (dst = (H5T_t *)H5I_object(dst_id)))
+            if (NULL == src || NULL == dst)
                 HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a datatype");
 
             src_parent = src->shared->parent;
 
             if (NULL == (tpath = H5T_path_find(src_parent, dst))) {
-                HGOTO_ERROR(H5E_DATASET, H5E_UNSUPPORTED, FAIL,
+                HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
                             "unable to convert between src and dest datatype");
             }
             else if (!H5T_path_noop(tpath)) {
-                if ((src_parent_id = H5I_register(H5I_DATATYPE, H5T_copy(src_parent, H5T_COPY_ALL), false)) <
-                    0)
-                    HGOTO_ERROR(H5E_DATASET, H5E_CANTREGISTER, FAIL,
-                                "unable to register types for conversion");
+                if (NULL == (tmp_type = H5T_copy(src_parent, H5T_COPY_ALL)))
```

Comment:
Why does this type need to be copied?   Same question for *tons* of other copies - I think the only reason 90% of them were originally copied was because they needed an ID and decrementing the ID would delete the type object.   Since we don't need the IDs, we shouldn't need the copies.

Reply:
Agreed (and added a comment above after thinking for a bit). My goal would be to tackle that directly after this PR, I just didn't want to dive into that yet since it can be tricky and this is also already a fairly large set of changes.